src/pool/srv_pool_scrub.c

```diff
@@ -4,324 +4,771 @@
+			DP_UUID((ctx)->sc_pool_uuid),
+			(ctx)->sc_dmi->dmi_tgt_id,
+			DP_UUID(ctx->sc_cont.scs_cont_uuid));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_start,
+			D_TM_TIMESTAMP,
+			"When the current scrubbing started", NULL,
+			DF_CONT_DIR"/"M_STARTED, DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_csum_calcs,
+			D_TM_COUNTER, "Number of checksums calculated for "
+				      "current scan",
+			NULL,
+			DF_CONT_DIR"/"M_CSUM_COUNTER, DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_last_csum_calcs,
+			D_TM_COUNTER, "Number of checksums calculated in last "
+				      "scan", NULL,
+			DF_CONT_DIR"/"M_CSUM_PREV_COUNTER,
+			DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_total_csum_calcs,
+			D_TM_COUNTER, "Total number of checksums calculated",
+			NULL,
+			DF_CONT_DIR"/"M_CSUM_TOTAL_COUNTER, DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_corruption,
+			D_TM_COUNTER, "Number of silent data corruption "
+				      "detected during current scan",
+			NULL,
+			DF_CONT_DIR"/"M_CSUM_CORRUPTION, DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_total_corruption,
+			D_TM_COUNTER, "Total number of silent data corruption "
+				      "detected",
+			NULL,
+			DF_CONT_DIR"/"M_CSUM_TOTAL_CORRUPTION,
+			DP_CONT_DIR(ctx));
+}
+
 static int
-scrub_cont(struct scrub_ctx *ctx)
+sc_scrub_cont(struct scrub_ctx *ctx)
 {
 	vos_iter_param_t		param = {0};
 	struct vos_iter_anchors		anchor = {0};
 	int				rc;
 
-	/** Do the actual scrubbing */
+	/* not all containers in the pool will have checksums enabled */
+	if (!daos_csummer_initialized(sc_csummer(ctx)))
+		return 0;
+
 	C_TRACE("Scrubbing container '"DF_UUIDF"'\n",
-		DP_UUID(ctx->cur_cont->sc_uuid));
+		DP_UUID(*sc_cont_uuid(ctx)));
 
-	param.ip_hdl = ctx->cur_cont->sc_hdl;
-	param.ip_epr.epr_lo = 0;
+	param.ip_hdl = sc_cont_hdl(ctx);
 	param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
-
 	rc = vos_iterate(&param, VOS_ITER_OBJ, true, &anchor,
-			 obj_iter_scrub_cb,
+			 obj_iter_scrub_pre_cb,
 			 NULL, ctx, NULL);
 
 	if (rc != DER_SUCCESS) {
 		D_ERROR("Object scrub failed: "DF_RC"\n", DP_RC(rc));
 		return rc;
 	}
 
+	ds_scrub_sched_control(ctx);
+
+	return 0;
+}
+
+static int
+sc_cont_setup(struct scrub_ctx *ctx, vos_iter_param_t *param,
+	      vos_iter_entry_t *entry)
+{
+	int			 rc;
+
+	if (ctx->sc_cont_lookup_fn == NULL)
+		return -DER_NOSYS;
+	rc = ctx->sc_cont_lookup_fn(ctx->sc_pool_uuid, entry->ie_couuid,
+				    ctx->sc_sched_arg, &ctx->sc_cont);
+
+	if (rc != 0) {
+		D_ERROR("Error opening vos container: "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
 	return 0;
 }
 
+static void
+sc_cont_teardown(struct scrub_ctx *ctx)
+{
+	memset(&ctx->sc_cont, 0, sizeof(ctx->sc_cont));
+}
+
 /** vos_iter_cb_t */
 static int
 cont_iter_scrub_cb(daos_handle_t ih, vos_iter_entry_t *entry,
 		   vos_iter_type_t type, vos_iter_param_t *param,
 		   void *cb_arg, unsigned int *acts)
```

Comment:
vos_cont_destroy will return DER_BUSY because the scrubber has a reference to the cont. I just made some changes that will allow the scrubber to know that the container is being stopped so it won't continue scrubbing.